storage/pin/pin.go

```diff
@@ -0,0 +1,408 @@
+// or collection. It also display three vital information's
+//     1) Whether the file is a RAW file or not
+//     2) Size of the pinned file or collection
+//     3) the number of times that particular file or collection is pinned.
+func (p *API) ListPinFiles() map[string]FileInfo {
+	p.pinLock.RLock()
+	defer p.pinLock.RUnlock()
+	for hash, fileInfo := range p.pinnedFiles {
+		log.Info("Pinned file", "Address", hash, "IsRAW", fileInfo.IsRaw,
+			"FileSize", fileInfo.FileSize, "PinCounter", fileInfo.FileSize)
+	}
+	return p.pinnedFiles
+}
+
+func (p *API) walkChunksFromRootHash(rootHash string, isRaw bool, credentials string, executeFunc func(storage.Reference) error) {
+	fileWorkers := make(chan storage.Reference, WorkerChanSize)
+	chunkWorkers := make(chan storage.Reference, WorkerChanSize)
+
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash", "err", err)
+		return
+	}
+
+	hashFunc := storage.MakeHashFunc(storage.DefaultHash)
+	hashSize := len(addr)
+	isEncrypted := len(addr) > hashFunc().Size()
+	getter := storage.NewHasherStore(p.db, hashFunc, isEncrypted, chunk.NewTag(0, "show-chunks-tag", 0))
+
+	go func() {
+		if !isRaw {
+
+			// If it is not a raw file... load the manifest and process the files inside one by one
+			walker, err := p.api.NewManifestWalker(context.Background(), storage.Address(addr),
+				p.api.Decryptor(context.Background(), credentials), nil)
+
+			if err != nil {
+				log.Error("Could not decode manifest.", "err", err)
+				return
+			}
+
+			err = walker.Walk(func(entry *api.ManifestEntry) error {
+
+				fileAddr, err := hex.DecodeString(entry.Hash)
+				if err != nil {
+					log.Error("Error decoding hash present in manifest", "err", err)
+					return err
+				}
+
+				// send the file to file workers
+				fileWorkers <- storage.Reference(fileAddr)
+
+				return nil
+			})
+
+			if err != nil {
+				log.Error("Error walking manifest", "err", err)
+				return
+			}
+
+			// Finally, remove the manifest file too
+			fileWorkers <- storage.Reference(addr)
+
+			// Signal end of file stream
+			close(fileWorkers)
+
+		} else {
+			// Its a raw file.. no manifest.. so process only this hash
+			fileWorkers <- storage.Reference(addr)
+
+			// Signal end of file stream
+			close(fileWorkers)
+
+		}
+	}()
+
+	for fileRef := range fileWorkers {
+
+		// Send the file to chunk workers
+		chunkWorkers <- fileRef
+
+		actualFileSize := uint64(0)
+		rcvdFileSize := uint64(0)
+		doneChunkWorker := make(chan struct{})
+		var cwg sync.WaitGroup // Wait group to wait for chunk processing to complete
+
+	QuitChunkFor:
+		for {
+			select {
+			case <-doneChunkWorker:
+				break QuitChunkFor
+
+			case ref := <-chunkWorkers:
+				cwg.Add(1)
+
+				go func() {
+
+					defer cwg.Done()
+
+					chunkData, err := getter.Get(context.Background(), ref)
+					if err != nil {
+						log.Error("Error getting chunk data from localstore.", "Address", hex.EncodeToString(ref))
```

Comment:
usually the walker function return an error. knowing that an error has occurred in this case will be very difficult to know. we'll have to some how flesh it out in some obscure logfile. i would be happy if the walker returns an error and that we don't get just a logline when there's an error.

Reply:
Refactored to emit error.